etag.go

```diff
@@ -0,0 +1,40 @@
+package main
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"net/http"
+)
+
+// checks whether client's ETag matches current response body.
+// - if the IMGPROXY_USE_ETAG env var is unset, this function always returns false
+// - if the IMGPROXY_USE_ETAG is set to "true", the function calculates current ETag and compares it
+//   with another ETag value provided by a client request
+// Note that the calculated ETag value is saved to outcoming response with "ETag" header.
+func isETagMatching(b []byte, po *processingOptions, rw *http.ResponseWriter, r *http.Request) bool {
+
+	if !conf.ETagEnabled {
+		return false
+	}
+
+	// calculate current ETag value using sha1 hashing function
+	currentEtagValue := calculateHashSumFor(b, po)
+	(*rw).Header().Set("ETag", currentEtagValue)
+	return currentEtagValue == r.Header.Get("If-None-Match")
+}
+
+// the function calculates the SHA checksum for the current image and current Processing Options.
+// The principal is very simple: if an original image is the same and POs are the same, then
+// the checksum must be always identical. But if PO has some different parameters, the
+// checksum must be different even if original images match
+func calculateHashSumFor(b []byte, po *processingOptions) string {
+	hash := sha1.New()
+
+	// SHA checksum consists of image, porcessing options and random value, generated on the server startup.
+	hash.Write(b)
+	binary.Write(hash, binary.LittleEndian, *po)
+	hash.Write(conf.RandomValue)
```

Comment:
I have noticed that you ignored one of my comments:

> I think this way to calc ETag is not perfectly correct.
>
> First of all, it messes image data with config, this may cause collisions.
> ...
>
> What should we do:
>
> 1. Calc SHA of image data. This way image footprint will have a constant length.
> 2. ...
> 3. ETag will be calculated like this: sha(image_footprint + po + cacheSignature)

```go
footprint := sha1.Sum(b)

hash := sha1.New()
hash.Write(footprint[:])
binary.Write(hash, binary.LittleEndian, *po)
hash.Write(conf.RandomValue)

return fmt.Sprintf("%x", hash.Sum(nil))
```

Reply:
why don't you change it by yourself? It is only one change of code, it is much faster and easier than writing these long comments on Github. The same question is regarding new lines and unnecessary comments deleting.

Reply:
Well, because it's you who wants this feature, not me. I appreciate your contribution, but I don't want to merge the code that doesn't meet my requirements. Because you won't maintain the project, I will. Ok, I could fix comments and blank lines things by myself. But I definitely won't merge feature that doesn't work properly. If you corrected my comments at once, the feature would be merged already, but you prefer "why can't you do my work yourself" thing. Come on.

Reply:
done :white_check_mark: